packages/plugins/trpc/src/generator.ts

```diff
@@ -89,35 +89,85 @@ function createAppRouter(
     const prismaImport = getPrismaClientImportSpec(zmodel, path.dirname(indexFile));
     appRouter.addImportDeclarations([
         {
-            namedImports: ['AnyRootConfig'],
+            namedImports: ['type AnyRootConfig', 'type Procedure', 'type ProcedureParams', 'type ProcedureType'],
             moduleSpecifier: '@trpc/server',
         },
         {
-            namedImports: ['PrismaClient'],
+            namedImports: ['type PrismaClient', 'type Prisma'],
             moduleSpecifier: prismaImport,
-            isTypeOnly: true,
         },
         {
-            namedImports: ['createRouterFactory', 'AnyRouter'],
+            namedImports: ['type createRouterFactory'],
             moduleSpecifier: '@trpc/server/dist/core/router',
         },
         {
-            namedImports: ['createBuilder'],
+            namedImports: ['type ProcedureBuilder'],
             moduleSpecifier: '@trpc/server/dist/core/internals/procedureBuilder',
         },
+        { defaultImport: 'z', moduleSpecifier: 'zod', isTypeOnly: true },
     ]);
 
     appRouter.addStatements(`
+        ${/** to be used by the other routers without making a bigger commit */ ''}
+        export { PrismaClient } from '${prismaImport}'; 
+
         export type BaseConfig = AnyRootConfig;
 
         export type RouterFactory<Config extends BaseConfig> = ReturnType<
             typeof createRouterFactory<Config>
         >;
         
-        export type ProcBuilder<Config extends BaseConfig> = ReturnType<
-            typeof createBuilder<Config>
-        >;
+        ${
+            /** this is needed in order to prevent type errors between a procedure and a middleware-extended procedure */ ''
+        }
+        export type ProcBuilder<Config extends BaseConfig> = ProcedureBuilder<{
+            _config: Config;
+            _ctx_out: Config['$types']['ctx'];
+            _input_in: any;
+            _input_out: any;
+            _output_in: any;
+            _output_out: any;
+            _meta: Config['$types']['meta'];
+        }>;
+
+        type ExtractParamsFromProcBuilder<Builder extends ProcedureBuilder<any>> =
+            Builder extends ProcedureBuilder<infer P> ? P : never;
         
+        type FromPromise<P extends Promise<any>> = P extends Promise<infer T>
+            ? T
+            : never;
+          
+        ${/** workaround to avoid creating 'typeof unsetMarker & object' on the procedure output */ ''}
+        type Join<A, B> = A extends symbol ? B : A & B;
+
+        ${
+            /** you can name it whatever you want, but this is to make sure that 
+                the types from the middleware and the procedure are correctly merged */ ''
+        }
+        export type ProcReturns<
```

Comment:
And this whole workaround is to "extract and merge" the types of I/O coming from the procedure with those from the schemas.

Reply:
Got it, thanks for the context! Let me look into why the tests are failing.